### PR TITLE
feat: add num_sessions_in_pool metric

### PR DESCRIPTION
**Pull request by mayurkale22**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-spanner/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Updates #53, this is based on the last discussion in the thread: https://github.com/googleapis/java-spanner/issues/53#issuecomment-601455550

This is how it looks in Stackdriver UI:
Screen Shot 2020-03-25 at 1 42 39 PM

**Breaking Note:**
The `cloud.google.com/java/spanner/in_use_sessions` metric has been deprecated since v1.55.2, use `cloud.google.com/java/spanner/num_sessions_in_pool` instead.

### CODE REVIEW

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/MetricRegistryConstants.java [hunk 1]
```diff
@@ -46,20 +55,20 @@
   // The Metric name and description
   static final String MAX_IN_USE_SESSIONS = "cloud.google.com/java/spanner/max_in_use_sessions";
   static final String MAX_ALLOWED_SESSIONS = "cloud.google.com/java/spanner/max_allowed_sessions";
-  static final String IN_USE_SESSIONS = "cloud.google.com/java/spanner/in_use_sessions";
```
Comment (skuruppu) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/MetricRegistryConstants.java]: I'm worried for anyone relying on this metric already in their monitoring? What is the process for deprecating metrics in OpenCensus? Is there some monitoring to figure out if someone is exporting it?

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java [hunk 2]
```diff
@@ -1638,7 +1637,6 @@ public Void call() {
         .containsEntry(MetricRegistryConstants.NUM_ACQUIRED_SESSIONS, 3L);
     assertThat(record.getMetrics())
         .containsEntry(MetricRegistryConstants.NUM_RELEASED_SESSIONS, 3L);
-    assertThat(record.getMetrics()).containsEntry(MetricRegistryConstants.IN_USE_SESSIONS, 0L);
```
Comment (rghetia) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java]: No test for new Gauges?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/MetricRegistryConstants.java [hunk 3]
```diff
@@ -30,12 +30,21 @@
...
+  private static final LabelKey SESSION_TYPE = LabelKey.create("Type", "Type of the Sessions");
 
   /** The label value is used to represent missing value. */
   private static final LabelValue UNSET_LABEL = LabelValue.create(null);
 
+  static final LabelValue NUM_IN_USE_SESSIONS = LabelValue.create("num_in_use_sessions");
+  static final LabelValue NUM_SESSIONS_BEING_PREPARED =
+      LabelValue.create("num_sessions_being_prepared");
+  static final LabelValue NUM_READ_SESSIONS = LabelValue.create("num_read_sessions");
+  static final LabelValue NUM_WRITE_SESSIONS = LabelValue.create("num_write_prepared_sessions");
```
[the remark this answers is not in the document]
Comment (mayurkale22) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/MetricRegistryConstants.java]: @skuruppu what do you suggest?

[the remark this answers is not in the document]
Comment (olavloite) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/MetricRegistryConstants.java]: Regarding `num_write_prepared_session` vs read sessions not having the word prepared: That is intentional. Sessions themselves are not read-only or read/write, but a portion of the sessions in the pool is prepared with a read/write transaction for direct use. So if a user application wants to execute a read/write transaction, the pool returns a session that already has a prepared read/write transaction associated with it.

**Comment by hengfengli** [comment]
@mayurkale22 I just want to check what is latest status of this PR? any progress?

**Comment by mayurkale22** [comment]
> @mayurkale22 I just want to check what is latest status of this PR? any progress?

Oops! This went off my radar due to other priorities. I'm going to rebase the PR and fix unresolved comments soon.

Thanks for bringing this to my attention.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java (hunk 2 above)
Comment (mayurkale22) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java:1760]: Added test to validate newly added Metric. I had to modify `MetricRegistryTestUtils` to support and test multiple timeseries entries. PTAL when you get a chance.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/MetricRegistryConstants.java (hunk 1 above)
Comment (mayurkale22) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/MetricRegistryConstants.java:49]: > What is the process for deprecating metrics in OpenCensus?

In OC, (usually) we first mark the existing API as deprecated and keep it for 18 months before removing it. I think the feature is fairly new and looks safe to remove it with a note in the Changelog.

> Is there some monitoring to figure out if someone is exporting it? 

I think it is feasible to see the ingested metrics in stackdriver, but I haven't done it personally.

> I'm worried for anyone relying on this metric already in their monitoring?

Another option is keep existing(`IN_USE_SESSIONS`) metric along with new metric. In this case, in_use_sessions data will be accounted/exported twice.

WDYT?

Comment (skuruppu) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/MetricRegistryConstants.java:49]: Ok, I think you're right that this is relatively new. We will mark this in the Changelog.

Comment (skuruppu) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/MetricRegistryConstants.java:49]: If you can include something about this in the PR description, that would be great. Then we can copy it in when doing the release.

Comment (mayurkale22) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/MetricRegistryConstants.java:49]: done

**Comment by mayurkale22** [comment]
@skuruppu @olavloite do you know why build is failing (Kokoro - Test: Integration), looks unrelated to my change.

**Comment by olavloite** [comment]
@mayurkale22 That is unfortunately an (infrastructure) problem that we are running into on a semi-regular basis. It has nothing to do with your change, and the only way to circumvent it at the moment is triggering a new build, for example by adding the kokoro:force-run label to the PR.

**Comment by skuruppu** [comment]
@olavloite, the [latest failure](https://source.cloud.google.com/results/invocations/28301b95-b463-4ffd-afe4-f66a33f3ad7a/targets/github%2Fjava-spanner%2Fgoogle-cloud-spanner%2Ftarget%2Ffailsafe-reports%2Fintegration%2Fcom.google.cloud.spanner.it.ITQueryOptionsTest/tests) is interesting:

```
com.google.cloud.spanner.SpannerException: INVALID_ARGUMENT: com.google.api.gax.rpc.InvalidArgumentException: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: Query optimizer version: 10000 is not supported
```

would you be able to take a quick look?

**Comment by olavloite** [comment]
> @olavloite, the [latest failure](https://source.cloud.google.com/results/invocations/28301b95-b463-4ffd-afe4-f66a33f3ad7a/targets/github%2Fjava-spanner%2Fgoogle-cloud-spanner%2Ftarget%2Ffailsafe-reports%2Fintegration%2Fcom.google.cloud.spanner.it.ITQueryOptionsTest/tests) is interesting:
> 
> ```
> com.google.cloud.spanner.SpannerException: INVALID_ARGUMENT: com.google.api.gax.rpc.InvalidArgumentException: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: Query optimizer version: 10000 is not supported
> ```
> 
> would you be able to take a quick look?

Hmmm... That's strange. There has been no changes to the test case lately, so my initial guess would be that the backend is now reacting differently. Up until now the backend accepted query options to be specified for DML statements, but seemed to ignore them. Now it seems that they are at least checked for a valid value.

**Comment by skuruppu** [comment]
That's interesting. I also wondered why this suddenly came up given that we haven't touched this code in a couple of months.

**Comment by olavloite** [comment]
> That's interesting. I also wondered why this suddenly came up given that we haven't touched this code in a couple of months.

I think we should just remove this specific test case for now, and ask the backend team what the expected behavior of `QueryOptions` should be for DML statements.

**Pull request merged**